opengovsg/plumber
# PLU-354: prevent batched queries in operations

## Problem

We previously disabled batching multiple operations in a single HTTP request. However, we still allow multiple root level fields in a single operation. It was flagged out during VAPT that this might facilitate potential DOS attacks and should only be enabled when strictly necessary.

## Solution

Create a custom apollo plugin to detect if there are multiple root level fields. If there are, throw a `BadUserInputError`.

## How to test

1. Run dev server and open http://localhost:3000/graphql to access the sandbox
2. Test with the following query 
```
query Batch {
  h1: healthcheck {
    version
  }
  h2: healthcheck {
    version
  }
}
```
3. You should be able to see the relevant error.
